microcosm_eventsource/event_types.py

```diff
@@ -195,6 +203,52 @@ def all_states(cls):
                     next_states.append(new_state)
                     known_states.add(new_state)
 
+    @classmethod
+    def all_states_and_events(cls):
+        """
+        Return a generator of all allowed (state, event_type) combinations
+        Note: it can return the same state or event twice (but all uniqu)
+
+        """
+        initial_states = list(cls.initial_states())
+        initial_states_and_events = [(state, list(state)[0]) for state in initial_states]
+        yield from initial_states_and_events
+        next_states = initial_states
+        known_states = set(initial_states)
+        known_states_and_events = set(initial_states_and_events)
+
+        while True:
+            if not next_states:
+                return
+            current_states = next_states
+            next_states = []
+            for state in current_states:
+                for event_type in cls.available_transitions(state):
+                    new_state = frozenset(event_type.accumulate_state(state))
+                    if (new_state, event_type) in known_states_and_events:
+                        continue
+                    known_states_and_events.add((new_state, event_type))
+                    yield new_state, event_type
+                    if new_state in known_states:
+                        continue
+                    next_states.append(new_state)
+                    known_states.add(new_state)
+
+    @classmethod
+    def all_transitions(cls, states=None):
+        """
+        Return a generator of all allowed transitions as a tuple of (initial state, new state, event type).
+
+        :param state: a list states to check, If None - all allowed states
+
+        """
+        if states is None:
+            states = cls.all_states()
+        for state in states:
+            for event_type in cls.available_transitions(state):
```

Comment:
We're actually calculating the transitions twice - but I prefer to save few lines